8331497: Implement JEP 483: Ahead-of-Time Class Loading & Linking

iklam opened the pull request:
This is an implementation of [JEP 483: Ahead-of-Time Class Loading & Linking](https://openjdk.org/jeps/483).

----
Note: this is a combined PR of the following individual PRs
- https://github.com/openjdk/jdk/pull/20516
- https://github.com/openjdk/jdk/pull/20517
- https://github.com/openjdk/jdk/pull/20843
- https://github.com/openjdk/jdk/pull/20958
- https://github.com/openjdk/jdk/pull/20959
- https://github.com/openjdk/jdk/pull/21049
- https://github.com/openjdk/jdk/pull/21143

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331497](https://bugs.openjdk.org/browse/JDK-8331497): Implement JEP 483: Ahead-of-Time Class Loading &amp; Linking (**Sub-task** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [f0bc1ae6](https://git.openjdk.org/jdk/pull/21642/files/f0bc1ae60fea80d5914d520457986a753e75fae4)
 * [John R Rose](https://openjdk.org/census#jrose) (@rose00 - **Reviewer**)
 * [Dan Heidinga](https://openjdk.org/census#heidinga) (@DanHeidinga - no project role) Review applies to [1b6c3e28](https://git.openjdk.org/jdk/pull/21642/files/1b6c3e288acb5e649d01fb73c9caaa2b6b0f3f48)
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer) Review applies to [935dcc68](https://git.openjdk.org/jdk/pull/21642/files/935dcc687d8fc13c591d0ef4c7428d55e37dde08)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) Review applies to [c46ae0b9](https://git.openjdk.org/jdk/pull/21642/files/c46ae0b913028978dbe7b17491b1bfc80dcd4dac)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21642/head:pull/21642` \
`$ git checkout pull/21642`

Update a local copy of the PR: \
`$ git checkout pull/21642` \
`$ git pull https://git.openjdk.org/jdk.git pull/21642/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21642`

View PR using the GUI difftool: \
`$ git pr show -t 21642`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21642.diff">https://git.openjdk.org/jdk/pull/21642.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21642#issuecomment-2429755620)
</details>

vnkozlov reviewed (approved):
Approved.

liach reviewed (commented):
The java file changes look good.

rose00 commented:
I took a closer look at the diffs in a file you renamed, from `classPrelinker.cpp` to `aotConstantPoolResolver.cpp`.  Changes still look good, notably the filtering logic on the bootstrap methods (and their arguments) that we support.

For the record, the diff I made for review is enclosed.  It makes the new indy-related logic stand out pretty well.

[classPrelinker.deleted.cpp.txt](https://github.com/user-attachments/files/17501832/classPrelinker.deleted.cpp.txt)

rose00 commented:
A thought for a possible cleanup, after this PR is done…

The scratch mirror logic had me… scratching my head.  It seems to me that a more descriptive name would make the code explain itself better.  I suggest (for a future cleanup) calling a mirror structure which is being aot-assembled (just for the archive) a "future mirror" (or maybe "production mirror" or "mirror asset").  This is in distinction to the current "live" mirror, which is also the AOT phase mirror.  In general, from the point of view of the assembly phase, when we build new structures not created by the JVM as a result of post-training Java execution, we might want to give them a common descriptive term.  But I suppose most items that make it into the AOT cache are faithful copies of live data, present in the VM at dump time (end of assembly phase).  In that case, it's more like "the same structure" in all cases, and there's no need to simultaneously work on both present and future versions of the same structure.

(When I say "structure" I mean mirror object for now, but perhaps the pattern might expand to something else?  Or, maybe we will get rid of the two-mirror solution, in which case every future structure is also completely present and live in the assembly-phase VM.)

rose00 reviewed (approved):
I read it through again.  Some stylistic issues we have discussed can be saved for later.  I didn't find any new problems, and known old problems are addressed. Ship it!

DanHeidinga reviewed (approved):
Looks good!

ashu-mehra commented on src/hotspot/share/cds/aotClassLinker.cpp:
```diff
@@ -0,0 +1,315 @@
+}
+
+void AOTClassLinker::add_candidate(InstanceKlass* ik) {
+  _candidates->put_when_absent(ik, true);
```
I just noticed the use of `put_when_absent` here. This means the caller should ensure that `ik` has not already been added to the `_candidates`. We do that currently (`try_add_candidate` calls `is_candidate` before calling `add_candidate`) but probably worth mentioning this contract explicitly in a comment somewhere for future readers.

iklam replied:
I renamed it to `add_new_candidate()` and added comments so that it's clear what the contract is. I also added assert to check for thread safety.

ashu-mehra commented:
@iklam I remember there was [ConstantPool::iterate_archivable_resolved_references](https://github.com/iklam/jdk/blob/f0bc1ae60fea80d5914d520457986a753e75fae4/src/hotspot/share/oops/constantPool.cpp#L288) in https://github.com/openjdk/jdk/pull/21143 but I don't see it any more in this PR. Can you please comment on why was that removed?

turbanoff commented on test/hotspot/jtreg/runtime/cds/appcds/aotClassLinking/AOTClassLinkingVMOptions.java:
```diff
@@ -0,0 +1,165 @@
+        return appClasses;
+    }
+
+    static void init() throws Exception  {
```
```suggestion
    static void init() throws Exception {
```

iklam commented:
> A thought for a possible cleanup, after this PR is done…
> 
> The scratch mirror logic had me… scratching my head. It seems to me that a more descriptive name would make the code explain itself better. I suggest (for a future cleanup) calling a mirror structure which is being aot-assembled (just for the archive) a "future mirror" (or maybe "production mirror" or "mirror asset"). This is in distinction to the current "live" mirror, which is also the AOT phase mirror. In general, from the point of view of the assembly phase, when we build new structures not created by the JVM as a result of post-training Java execution, we might want to give them a common descriptive term. But I suppose most items that make it into the AOT cache are faithful copies of live data, present in the VM at dump time (end of assembly phase). In that case, it's more like "the same structure" in all cases, and there's no need to simultaneously work on both present and future versions of the same structure.
> 
> (When I say "structure" I mean mirror object for now, but perhaps the pattern might expand to something else? Or, maybe we will get rid of the two-mirror solution, in which case every future structure is also completely present and live in the assembly-phase VM.)

The cached heap objects are mostly copied as-is, with a recursive walk from a set of roots. However, in some cases, we need to perform transformation in some of the objects. The transformation is implemented by substituting some of the discovered objects with a "scratch" (or "future") version.

For example, for java mirrors:

- If a class K1 *is not* aot-initialized, we need to zero out most of the fields inside `K1->java_mirror()`, but keep the injected `klass` and `array_klass` native pointers.
- If a class K2 *is* aot-initialized, we need to also keep the static fields declared in Java code in `K2->java_mirror()`

For example, here are the contents of the aot-cached mirror for the java/lang/String class:

```
 - ---- fields (total size 17 words):
 - private volatile transient 'classRedefinedCount' 'I' @12  0 (0x00000000)
 - injected 'klass' 'J' @16  2684621920 (0x00000000a0041460)
 - injected 'array_klass' 'J' @24  2684707368 (0x00000000a0056228)
 - injected 'oop_size' 'I' @32  17 (0x00000011)
 - injected 'static_oop_field_count' 'I' @36  2 (0x00000002)
 - private volatile transient 'cachedConstructor' 'Ljava/lang/reflect/Constructor;' @40  null (0x00000000)
 - private transient 'name' 'Ljava/lang/String;' @44  null (0x00000000)
 - private transient 'module' 'Ljava/lang/Module;' @48  null (0x00000000)
 - private final 'classLoader' 'Ljava/lang/ClassLoader;' @52  null (0x00000000)
 - private transient 'classData' 'Ljava/lang/Object;' @56  null (0x00000000)
 - private transient 'signers' '[Ljava/lang/Object;' @60  null (0x00000000)
 - private transient 'packageName' 'Ljava/lang/String;' @64  null (0x00000000)
 - private final 'componentType' 'Ljava/lang/Class;' @68  null (0x00000000)
 - private volatile transient 'reflectionData' 'Ljava/lang/ref/SoftReference;' @72  null (0x00000000)
 - private volatile transient 'genericInfo' 'Lsun/reflect/generics/repository/ClassRepository;' @76  null (0x00000000)
 - private volatile transient 'enumConstants' '[Ljava/lang/Object;' @80  null (0x00000000)
 - private volatile transient 'enumConstantDirectory' 'Ljava/util/Map;' @84  null (0x00000000)
 - private volatile transient 'annotationData' 'Ljava/lang/Class$AnnotationData;' @88  null (0x00000000)
 - private volatile transient 'annotationType' 'Lsun/reflect/annotation/AnnotationType;' @92  null (0x00000000)
 - transient 'classValueMap' 'Ljava/lang/ClassValue$ClassValueMap;' @96  null (0x00000000)
 - injected 'protection_domain' 'Ljava/lang/Object;' @100  null (0x00000000)
 - injected 'source_file' 'Ljava/lang/Object;' @104  null (0x00000000)
 - injected '<init_lock>' 'Ljava/lang/Object;' @108  [I{0x000000060ec016b0} (0xc1d802d6)
 - signature: Ljava/lang/String;
 - ---- static fields (2):
 - private static final 'serialVersionUID' 'J' @120  -6849794470754667710 (0xa0f0a4387a3bb342)
 - static final 'COMPACT_STRINGS' 'Z' @130  true (0x01)
 - private static final 'serialPersistentFields' '[Ljava/io/ObjectStreamField;' @112  a 'java/io/ObjectStreamField'[0] {0x000000060ec0e468} (0xc1d81c8d)
 - private static final 'REPL' 'C' @128    65533 (0xfffd)
 - public static final 'CASE_INSENSITIVE_ORDER' 'Ljava/util/Comparator;' @116  a 'java/lang/String$CaseInsensitiveComparator'{0x000000060ec0e7a8} (0xc1d81cf5)
 - static final 'LATIN1' 'B' @131  0 (0x00)
 - static final 'UTF16' 'B' @132  1 (0x01)
```

Here's a matrix for deciding when a field is kept or zeroed out:

```
Field              Type                     Not-AOT-inited   AOT-inited
-------------------------------------------------------------------------                               
"klass"            injected by HotSpot      keep             keep
"COMPACT_STRINGS"  declared in String.java  zero             keep
"module"           field of j.l.Class       zero             zero
```

After JEP 483 is integrated, I will investigate if the transformation can be done inside the recursive copying code without the substitution of scratch (future) objects.

An alternative to substitution would be to modify the original Java mirrors (to zero out fields that we don't want to cache), but that will damage the state of the VM. That's something CDS has been avoiding doing -- we want the VM to be useable after the AOT cache is written.

Of course, the holy grail is to avoid any transformation and simply cache everything as-is, in a snapshot style. We can't do that yet as the AOT cache still requires some "loading" operations in the production run. Some of the loading code, for example, might be confused if it sees `module` to be non-null when an aot-inited class is loaded into the JVM.

iklam commented:
> @iklam I remember there was [ConstantPool::iterate_archivable_resolved_references](https://github.com/iklam/jdk/blob/f0bc1ae60fea80d5914d520457986a753e75fae4/src/hotspot/share/oops/constantPool.cpp#L288) in #21143 but I don't see it any more in this PR. Can you please comment on why was that removed?

An indy call site has two entries in the resolved_references array:
- (a) The resolved call site
- (b) The MethodHandle of the boot strap method

We weren't archiving (b), which caused some JIT replay code to fail (the JIT assumes that both (a) and (b) must be resolved together). To fix this problem, I added the archiving of (b) in https://github.com/openjdk/jdk/pull/21642/commits/1d3daa4299691e75450d56ed1608b77e75cfd00a

Afterwards, I realized that we can simply archive everything in the resolved_reference array, because AOTConstantPoolResolver::is_indy_resolution_deterministic() already ensures that we only resolve indys that we can support. `ConstantPool::iterate_archivable_resolved_references()` is no longer necessary as it simply repeats the checks in `is_indy_resolution_deterministic()`. That's why I removed it in 
https://github.com/openjdk/jdk/pull/21642/commits/1d3daa4299691e75450d56ed1608b77e75cfd00a

But, now I realized that this part is also removed, which is probably not good.

```
    Array<ResolvedMethodEntry>* method_entries = cache()->resolved_method_entries();
    if (method_entries != nullptr) {
      for (int i = 0; i < method_entries->length(); i++) {
        ResolvedMethodEntry* rme = method_entries->adr_at(i);
        if (rme->is_resolved(Bytecodes::_invokehandle) && rme->has_appendix() &&
            cache()->can_archive_resolved_method(this, rme)) {
          int rr_index = rme->resolved_references_index();
          assert(resolved_reference_at(rr_index) != nullptr, "must exist");
          function(rr_index);
        }
      }
    }
```

For safety, I will revert the removal. We can consider refactoring the code after 483 is integrated.

iklam commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    if (CDSConfig::is_dumping_aot_linked_classes()) {
+      // Need to call try_add_candidate instead of is_candidate, as this may be called
```
In the future, these functions may be called before we enter the safepoint (e.g., to check if some constant pool entries can be resolved), so I think it's better to keep the `try_add_candidate()` call.

iklam commented on src/hotspot/share/cds/heapShared.hpp:
```diff
@@ -282,12 +295,14 @@ class HeapShared: AllStatic {
 
-  // The "default subgraph" is the root of all archived objects that do not belong to any
-  // of the classes defined in the <xxx>_archive_subgraph_entry_fields[] arrays:
+  // The "special subgraph" contains all the all archived objects that are reachable
```
Fixed.

iklam commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -629,7 +646,79 @@ class UnregisteredClassesDuplicationChecker : StackObj {
+  } else {
+    InstanceKlass* ik = InstanceKlass::cast(k);
+
+    if (SafepointSynchronize::is_at_safepoint()) {
```
I renamed the public function to `should_be_excluded(InstanceKlass*)` to avoid confusion with the private function `check_for_exclusion(InstanceKlass*, DumpTimeClassInfo*)`. I also added code to make sure that exclusion has been checked for the class, even when in safepoint.

DanHeidinga commented on src/hotspot/share/cds/aotClassInitializer.cpp:
```diff
@@ -0,0 +1,353 @@
+      log_info(cds, init)("Calling %s::runtimeSetup()", ik->external_name());
+    }
+    JavaValue result(T_VOID);
+    JavaCalls::call_static(&result, ik,
```
Based on the discussions in JDK-8342283, do we need a memory fence after the call to runtimeSetup?

I think we do to be consistent with the non-AOTCache run as there is a fence after `<clinit>` (when this code would normally run) due to the initialization lock.

iklam replied:
I think this is already done:

`runtimeSetup()` is called inside `InstanceKlass::call_class_initializer()`. When that returns, we will proceed to `InstanceKlass::set_initialization_state_and_notify()` which will perform the memory fencing. 

https://github.com/iklam/jdk/blob/6eebd18fc2820ffb179d431f179fc6af6d1be247/src/hotspot/share/oops/instanceKlass.cpp#L1276-L1300

DanHeidinga replied:
I think you're right about the memory fencing being covered there but reading that code shows another potential issue - what if the `runtimeSetup()` call throws an exception?

If I read `InstanceKlass::call_class_initializer()` correctly, an exception from `runtimeSetup()` will lead to marking the class as failing initialization (`initialization_error`) which will be an unexpected outcome as there may be instances of the class in the AOTCache that aren't produced by paths through the `<clinit>` method.

Classes are able to create an instance in `<clinit>` and escape it to other code and then fail initialization.  The instances are left in an odd state where after the `<clinit>` the instance methods can be called but static fields and static methods can't be accessed.

`runtimeSetup()` expands on this corner case which I think is OK for current "crawl" stage of our "crawl-walk-run" approach as its analogous to the existing `<clinit>` behaviour.  It may need refining in the future to abort the VM or do something else (?) in the case of failed `runtimeSetup` calls

iklam replied:
Thanks for catching this. I added code to exit the JVM when `runtimeSetup()` fails.

iwanowww reviewed (approved):
Looks good.

Stylistic comment: while browsing the code a mix of "AOT" & "aot" caught my eye. I find the former spelling more descriptive. Any particular reason to use "aot" unless all lowercase spelling is preferred in some particular context?

iklam commented:
> Looks good.
> 
> Stylistic comment: while browsing the code a mix of "AOT" & "aot" caught my eye. I find the former spelling more descriptive. Any particular reason to use "aot" unless all lowercase spelling is preferred in some particular context?

I've been using AOT in class names and aot in function/variable names. I think that's consistent with HotSpot convention.

In logging, I've been using "aot" as logging seems to favor lowercase.

In comments, I've been using a mix of AOT-xxx and aot-xxx, for no particular reason :-(

dholmes-ora commented:
> I've been using AOT in class names and aot in function/variable names. I think that's consistent with HotSpot convention.

As an acronym I think AOT in capitals should be preferred even in function/variable names. 

> In logging, I've been using "aot" as logging seems to favor lowercase.

Yes log tags seem to be always lower-case, by convention.

iklam commented:
> > I've been using AOT in class names and aot in function/variable names. I think that's consistent with HotSpot convention.
> 
> As an acronym I think AOT in capitals should be preferred even in function/variable names.

I searched for "GC" vs "gc" and "JVMTI" vs "jvmti" in function names. Almost all of them are in lower cases.

coleenp commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -0,0 +1,573 @@
+/*
+ * Copyright (c) 2022, 2024, Oracle and/or its affiliates. All rights reserved.
```
If this is a new file, it shouldn't have 2022 in the copyright year.

coleenp replied:
Oh looks like it's been moved from classPrelinker, except git doesn't show it's moved.  If it's added as a moved file, would there be less diffs to look at?

coleenp commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -0,0 +1,573 @@
+      return false;
+    }
+
+    if (!k->is_instance_klass()) {
```
How can a field, interface or method cp reference point to a non-instance klass?  Or is this for a method in java.lang.Object, which can be an array klass ?  I doubt you should bother to support this special case.

coleenp commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -0,0 +1,573 @@
+      return is_class_resolution_deterministic(cp_holder, InstanceKlass::cast(elem));
+    } else if (elem->is_typeArray_klass()) {
+      return true;
+    } else {
```
I think bottom_klass can only be InstanceKlass or typeArrayKlass, so else should be not possible and an assert.

coleenp reviewed (commented):
I had a couple of drive by comments.

rose00 reviewed (approved):
Reviewed again.

iklam commented:
Thanks everyone for the review and support!
I ran tiers 1~8 several times in the past few days when merging with mainline. No regressions found.
GHA error is timeout and seems to be unrelated. Integrating now.
/integrate

magicus commented on make/RunTests.gmk:
```diff
@@ -702,6 +703,58 @@ define SetJtregValue
   endif
 endef
 
+
```
This PR was not tagged with the `build` label, and was therefore not brought to the attention of the build team.

This file contains a large number of unacceptable changes. :-( It would under no circumstance have been approved as it looks. 

Please make sure that the Github labels actually matches the set of files that are changed in the PR. Skara tooling will assist in this, but the ultimate responsibility lies with the Author and Reviewers of the PR.
